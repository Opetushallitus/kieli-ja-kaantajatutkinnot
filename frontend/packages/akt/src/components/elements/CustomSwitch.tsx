import {
  FormControlLabel,
  FormGroup,
  Switch,
  SwitchProps,
} from '@mui/material';

import { Caption, Text } from 'components/elements/Text';
import { Color } from 'enums/app';

interface CustomSwitchProps extends SwitchProps {
  dataTestId?: string;
  leftLabel: string;
  rightLabel: string;
  errorLabel?: string;
  value?: boolean;
}

export const CustomSwitch = ({
  dataTestId,
  leftLabel,
  rightLabel,
  errorLabel,
  value,
  disabled,
  onChange,
}: CustomSwitchProps) => {
  const leftLabelClassName = disabled
    ? 'color-disabled margin-right-xs'
    : 'margin-right-xs';

  return (
    <FormGroup className="rows">
      <div className="columns">
        <Text className={leftLabelClassName}>{leftLabel}</Text>
        <FormControlLabel
          disabled={disabled}
          control={
            <Switch
              data-testid={dataTestId}
              checked={value}
              color={Color.Secondary}
              onChange={onChange}
            />
          }
          label={rightLabel}
        />
      </div>
      {errorLabel && (
        <Caption data-testid={`${dataTestId}__error-label`} color={Color.Error}>
          {errorLabel}
        </Caption>
      )}
    </FormGroup>
  );
};
