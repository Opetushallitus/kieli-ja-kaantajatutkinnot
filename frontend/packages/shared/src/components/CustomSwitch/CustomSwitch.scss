.custom-switch {
  &__helper-text {
    &#{&} {
      margin-left: 0.5rem;
    }
  }
}
