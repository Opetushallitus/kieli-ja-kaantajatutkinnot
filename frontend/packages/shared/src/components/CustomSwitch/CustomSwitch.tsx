import {
  FormControlLabel,
  FormGroup,
  Switch,
  SwitchProps,
} from '@mui/material';
import { ReactNode } from 'react';

import { Color } from '../../enums/common';
import { Caption, Text } from '../Text/Text';
import './CustomSwitch.scss';

interface CustomSwitchProps extends SwitchProps {
  dataTestId?: string;
  leftLabel: string;
  rightLabel: string;
  errorLabel?: ReactNode;
  value?: boolean;
}

export const CustomSwitch = ({
  dataTestId,
  leftLabel,
  rightLabel,
  errorLabel,
  value,
  disabled,
  onChange,
}: CustomSwitchProps) => {
  const leftLabelClassName = disabled
    ? 'color-disabled margin-right-xs'
    : 'margin-right-xs';

  return (
    <FormGroup className="rows">
      <div className="columns">
        <Text className={leftLabelClassName}>{leftLabel}</Text>
        <FormControlLabel
          className="custom-switch__label"
          disabled={disabled}
          control={
            <Switch
              data-testid={dataTestId}
              checked={value}
              color={Color.Secondary}
              onChange={onChange}
            />
          }
          label={rightLabel}
        />
      </div>
      {errorLabel && (
        <Caption
          className="custom-switch__helper-text"
          data-testid={`${dataTestId}__error-label`}
          color={'error'}
        >
          {errorLabel}
        </Caption>
      )}
    </FormGroup>
  );
};
