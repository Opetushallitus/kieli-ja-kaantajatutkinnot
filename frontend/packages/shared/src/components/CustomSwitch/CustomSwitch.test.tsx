import renderer from 'react-test-renderer';

import { CustomSwitch } from './CustomSwitch';

describe('CustomSwitch', () => {
  it('should render correctly with an error label', () => {
    const tree = renderer
      .create(
        <CustomSwitch
          dataTestId="just-a-test-id"
          leftLabel="leftLabel"
          rightLabel="rightLabel"
          errorLabel="test-error-label"
          value={true}
          disabled={true}
        ></CustomSwitch>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with just the minimal props', () => {
    const tree = renderer
      .create(
        <CustomSwitch
          leftLabel="anotherLeftLabel"
          rightLabel="anotherRightLabel"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
