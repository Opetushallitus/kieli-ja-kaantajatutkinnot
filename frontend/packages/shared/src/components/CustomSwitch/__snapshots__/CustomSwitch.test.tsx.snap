// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSwitch should render correctly with an error label 1`] = `
<div
  className="MuiFormGroup-root rows css-dmmspl-MuiFormGroup-root"
>
  <div
    className="columns"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 color-disabled margin-right-xs css-ahj2mt-MuiTypography-root"
    >
      leftLabel
    </p>
    <label
      className="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd custom-switch__label css-j204z7-MuiFormControlLabel-root"
    >
      <span
        className="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
      >
        <span
          aria-disabled={true}
          className="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary Mui-checked Mui-disabled Mui-disabled PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary Mui-checked Mui-disabled Mui-checked Mui-disabled css-ob1t9r-MuiButtonBase-root-MuiSwitch-switchBase"
          data-testid="just-a-test-id"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
            disabled={true}
            onChange={[Function]}
            required={false}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
          />
        </span>
        <span
          className="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
        />
      </span>
      <span
        className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled css-ahj2mt-MuiTypography-root"
      >
        rightLabel
      </span>
    </label>
  </div>
  <span
    className="MuiTypography-root MuiTypography-caption custom-switch__helper-text css-19m5e5t-MuiTypography-root"
    data-testid="just-a-test-id__error-label"
  >
    test-error-label
  </span>
</div>
`;

exports[`CustomSwitch should render correctly with just the minimal props 1`] = `
<div
  className="MuiFormGroup-root rows css-dmmspl-MuiFormGroup-root"
>
  <div
    className="columns"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 margin-right-xs css-ahj2mt-MuiTypography-root"
    >
      anotherLeftLabel
    </p>
    <label
      className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd custom-switch__label css-j204z7-MuiFormControlLabel-root"
    >
      <span
        className="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
      >
        <span
          className="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorSecondary css-ob1t9r-MuiButtonBase-root-MuiSwitch-switchBase"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <input
            className="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
            onChange={[Function]}
            required={false}
            type="checkbox"
          />
          <span
            className="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
          />
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          className="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
        />
      </span>
      <span
        className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        anotherRightLabel
      </span>
    </label>
  </div>
</div>
`;
